### ISSUE TITLE
[bug] Deploy generator messes up when you have two levels of symlinks

### ISSUE THREAD

**Issue by tru** [conan-io/conan#7104]
### Environment Details (include every applicable attribute)
  * Operating System+version: Linux
  * Compiler+version: n/a
  * Conan version: 1.25.2
  * Python version: 3.6.4

### Steps to reproduce (Include if Applicable)
Deploy a package where you have symlinks in two levels:
```
a->b
b->c
```

that broke because b was written before c was written and it tries to make a hash of it for the manifest:

### Logs (Executed commands with output) (Include/Attach if Applicable)

```
ERROR: Traceback (most recent call last):
  File "/Users/tobias/.conan_plex/lib/conan-1.25.2-102-darwin-py36.pyz/conans/client/generators/__init__.py", line 106, in write_generators
    content = generator.content
  File "/Users/tobias/.conan_plex/lib/conan-1.25.2-102-darwin-py36.pyz/conans/client/generators/deploy.py", line 56, in content
    return self.deploy_manifest_content(copied_files)
  File "/Users/tobias/.conan_plex/lib/conan-1.25.2-102-darwin-py36.pyz/conans/client/generators/deploy.py", line 22, in deploy_manifest_content
    file_dict[f] = md5sum(abs_path)
  File "/Users/tobias/.conan_plex/lib/conan-1.25.2-102-darwin-py36.pyz/conans/util/files.py", line 136, in md5sum
    return _generic_algorithm_sum(file_path, "md5")
  File "/Users/tobias/.conan_plex/lib/conan-1.25.2-102-darwin-py36.pyz/conans/util/files.py", line 149, in _generic_algorithm_sum
    with open(file_path, 'rb') as fh:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/tobias/Code/clang-builder/plex-llvm-toolchain/SDKs/MacOSX.sdk/usr/lib/libkrb5.tbd'
```

**Comment by jgsogo** [comment]
Hi! The manifest is computed after all files have been copied https://github.com/conan-io/conan/blob/41afc0e355caabdbc9d3ed42badb2b47dcbccdae/conans/client/generators/deploy.py#L56

So it has to be a bug while copying the files or assigning the symlinks. Maybe we have introduced a bug in this PR (https://github.com/conan-io/conan/pull/7044) merged into `1.25.2`. Probably the file they are linking to is not copied or it arrives to a different folder 🐛 

Can you share more details (paths inside the package) about the symlinks so I can reproduce the issue locally?

Thanks!

**Comment by tru** [comment]
```
➜ ls -al /Users/tobias/.conan_plex/.conan/data/apple-sdks/11.3.1/plex/toolchain/package/46f53f156846659bf39ad6675fa0ee8156e859fe/MacOSX.sdk/usr/lib/libkrb5.tbd
lrwxr-xr-x  1 tobias  staff  63 May 27 11:16 /Users/tobias/.conan_plex/.conan/data/apple-sdks/11.3.1/plex/toolchain/package/46f53f156846659bf39ad6675fa0ee8156e859fe/MacOSX.sdk/usr/lib/libkrb5.tbd -> ../../System/Library/Frameworks/Kerberos.framework/Kerberos.tbd
➜ ls -al /Users/tobias/.conan_plex/.conan/data/apple-sdks/11.3.1/plex/toolchain/package/46f53f156846659bf39ad6675fa0ee8156e859fe/MacOSX.sdk/usr/lib/../../System/Library/Frameworks/Kerberos.framework/Kerberos.tbd
lrwxr-xr-x  1 tobias  staff  29 May 27 11:17 /Users/tobias/.conan_plex/.conan/data/apple-sdks/11.3.1/plex/toolchain/package/46f53f156846659bf39ad6675fa0ee8156e859fe/MacOSX.sdk/usr/lib/../../System/Library/Frameworks/Kerberos.framework/Kerberos.tbd -> Versions/Current/Kerberos.tbd
➜ ls -la /Users/tobias/.conan_plex/.conan/data/apple-sdks/11.3.1/plex/toolchain/package/46f53f156846659bf39ad6675fa0ee8156e859fe/MacOSX.sdk/usr/lib/../../System/Library/Framework
s/Kerberos.framework/Versions/Current/Kerberos.tbd
-rw-r--r--  1 tobias  staff  28199 May 19 17:40 /Users/tobias/.conan_plex/.conan/data/apple-sdks/11.3.1/plex/toolchain/package/46f53f156846659bf39ad6675fa0ee8156e859fe/MacOSX.sdk/usr/lib/../../System/Library/Frameworks/Kerberos.framework/Versions/Current/Kerberos.tbd
```

hope this helps

**Comment by jgsogo** [comment]
Not failing here 😞 

```
⇒  conan --version
Conan version 1.25.2
```

```
⇒  ls -la /Users/jgsogo/.conan/data/name/version/_/_/package/103f6067a947f366ef91fc1b7da351c588d1827f/MacOSX.sdk/usr/lib/libkrb5.tbd 
lrwxr-xr-x  1 jgsogo  staff  63 29 may 09:45 /Users/jgsogo/.conan/data/name/version/_/_/package/103f6067a947f366ef91fc1b7da351c588d1827f/MacOSX.sdk/usr/lib/libkrb5.tbd -> ../../System/Library/Frameworks/Kerberos.framework/Kerberos.tbd
```

```
⇒  ls -la /Users/jgsogo/.conan/data/name/version/_/_/package/103f6067a947f366ef91fc1b7da351c588d1827f/MacOSX.sdk/usr/lib/../../System/Library/Frameworks/Kerberos.framework/Kerberos.tbd
lrwxr-xr-x  1 jgsogo  staff  29 29 may 09:45 /Users/jgsogo/.conan/data/name/version/_/_/package/103f6067a947f366ef91fc1b7da351c588d1827f/MacOSX.sdk/usr/lib/../../System/Library/Frameworks/Kerberos.framework/Kerberos.tbd -> Versions/Current/Kerberos.tbd
```

```
⇒  ls -la /Users/jgsogo/.conan/data/name/version/_/_/package/103f6067a947f366ef91fc1b7da351c588d1827f/MacOSX.sdk/usr/lib/../../System/Library/Frameworks/Kerberos.framework/Versions/Current/Kerberos.tbd 
-rw-r--r--  1 jgsogo  staff  0 29 may 09:39 /Users/jgsogo/.conan/data/name/version/_/_/package/103f6067a947f366ef91fc1b7da351c588d1827f/MacOSX.sdk/usr/lib/../../System/Library/Frameworks/Kerberos.framework/Versions/Current/Kerberos.tbd
```

Does it fail with other Conan versions?

**Comment by tru** [comment]
My guess is that it has to do with the order the files are copied - so that the final file is not there when the code tries to dereference the symlink and make the hash?

**Comment by jgsogo** [comment]
It shouldn't be the case. The hash is computed after all the files for all the dependencies have been copied. It is the last statement [here](https://github.com/conan-io/conan/blob/41afc0e355caabdbc9d3ed42badb2b47dcbccdae/conans/client/generators/deploy.py#L31).

Your failure is related to this file:

```
FileNotFoundError: [Errno 2] No such file or directory: '/Users/tobias/Code/clang-builder/plex-llvm-toolchain/SDKs/MacOSX.sdk/usr/lib/libkrb5.tbd'
```

It looks like it doesn't belong to the `apple-sdks` package you are deploying. Does the package contain a symlink to a file outside the package? Is this file in the filesystem? 

After #6922 symlinks are copied, but it will fail for broken symlinks (we need to fix that PR adding this use-case).

**Comment by memsharded** [comment]
Closing this ticket as outdated, Conan 2.X no longer has this deploy generator, and has completely new ``deployers`` feature.
